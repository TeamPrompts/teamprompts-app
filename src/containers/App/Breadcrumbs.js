import React from 'react';
import { Link } from 'react-router-dom';
import { BLANK_XS, pageTypes } from '../../constants';

function Breadcrumbs({ pageType, tag }) {
  switch (pageType) {
    case pageTypes.CategoryPage:
      return (
        <div className="mt-2">
          <Link className="text-indigo-700 underline" to="/">
            Home
          </Link>
          {' > '}
          <span className="capitalize">{tag.name}</span>
        </div>
      );
    case pageTypes.HomePage:
    default:
      return <div className="mt-2">&nbsp;</div>;
    case pageTypes.PromptsPage:
      return (
        <div className="mt-2">
          <Link className="text-indigo-700 underline" to="/">
            Home
          </Link>
          {' > '}
          <Link
            className="capitalize text-indigo-700 underline"
            to={`/${tag.slug}`}
          >
            {tag.name}
          </Link>
          {' > '}
          {BLANK_XS}
        </div>
      );
  }
}

export default Breadcrumbs;
