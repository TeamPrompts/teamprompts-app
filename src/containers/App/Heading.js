import React from 'react';

function Heading({ title }) {
  return (
    <h1 className="font-bold mb-8 mt-2 text-4xl text-gray-900">
      {title} <span className="antialiased text-base text-gray-600">Î²eta</span>
    </h1>
  );
}

export default Heading;
