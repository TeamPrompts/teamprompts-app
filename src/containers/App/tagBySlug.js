import tagAll from '../../api/tagAll';

export const tagEmpty = { name: '', slug: '' };

function tagBySlug({ slug, tags }) {
  if (tags.length > 0) {
    const tag = tags.find(tag => tag.slug === slug);
    return tag || tagAll;
  }
  return tagEmpty;
}

export default tagBySlug;
