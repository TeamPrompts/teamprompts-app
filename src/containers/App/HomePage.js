import React, { useEffect } from 'react';
import tagAll from '../../api/tagAll';
import Filters from '../../components/Filters/Filters';
import FillInTheBlanks from '../../FillInTheBlanks/FillInTheBlanks';
import { useAmplitude } from '../../instrumentation/AmplitudeHookProvider';
import { viewHome } from '../../instrumentation/events';
import Heading from './Heading';

function HomePage({ fitbs, history, match: { url }, tags }) {
  const { logEvent } = useAmplitude();
  const { properties, type } = viewHome;

  useEffect(() => {
    logEvent(type, properties({ url }));
  }, [logEvent, properties, type, url]);

  return (
    <>
      <Heading title="TeamPrompts" />
      <Filters filter={tagAll} fitbs={fitbs} tags={tags} />
      <ul>
        {fitbs.map((fitb, index) => (
          <li key={fitb.id}>
            <FillInTheBlanks
              fitb={fitb}
              history={history}
              viewPosition={index}
            />
          </li>
        ))}
      </ul>
    </>
  );
}

export default HomePage;
