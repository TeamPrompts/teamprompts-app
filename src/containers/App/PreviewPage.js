import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { ID } from '../../api/constants';
import selectFitbs from '../../api/selectFitbs';
import FillInTheBlanks from '../../FillInTheBlanks/FillInTheBlanks';
import Heading from './Heading';

function PreviewPage() {
  const { id } = useParams();

  const [error, setError] = useState();
  const [fitbs, setFitbs] = useState();
  const [waiting, setWaiting] = useState(true);

  useEffect(() => {
    selectFitbs({ filterByFormula: `${ID}="${id}"` }, (error, fitbs) => {
      if (error) {
        setError(error);
      } else {
        setFitbs(fitbs);
      }
    });
  }, [id]);

  useEffect(() => {
    if (error) {
      setWaiting(false);
    }
    if (fitbs) {
      setWaiting(false);
    }
  }, [error, fitbs]);

  return (
    <>
      <Heading title="TeamPrompts" />
      <ul>
        {!error &&
          !waiting &&
          fitbs.map(fitb => (
            <li key={fitb.id}>
              <FillInTheBlanks fitb={fitb} />
            </li>
          ))}
        {!error && !waiting && fitbs.length === 0 && (
          <li>
            <pre>404</pre>
          </li>
        )}
      </ul>
      {error && <pre>{JSON.stringify(error, 0, 2)}</pre>}
    </>
  );
}

export default PreviewPage;
