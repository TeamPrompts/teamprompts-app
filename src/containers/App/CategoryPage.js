import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import Filters from '../../components/Filters/Filters';
import FillInTheBlanks from '../../FillInTheBlanks/FillInTheBlanks';
import Heading from './Heading';
import tagBySlug from './tagBySlug';
import { useAmplitude } from '../../instrumentation/AmplitudeHookProvider';
import { viewCollection } from '../../instrumentation/events';

function fitbsByTag({ fitbs, tag, tags }) {
  return fitbs.filter(fitb => fitb.tags.includes(tag.id));
}

function CategoryPage({ fitbs, history, tags }) {
  const { slug } = useParams();
  const tag = tagBySlug({ slug, tags });

  const { logEvent } = useAmplitude();
  const { properties, type } = viewCollection;

  useEffect(() => {
    logEvent(type, properties({ counter: 'TBD', tag }));
  }, [logEvent, properties, tag, type]);

  return (
    <>
      <Heading title="TeamPrompts" />
      <Filters filter={{ name: tag.name }} fitbs={fitbs} tags={tags} />
      <ul>
        {fitbsByTag({ fitbs, tag, tags }).map(fitb => (
          <li key={fitb.id}>
            <FillInTheBlanks fitb={fitb} history={history} tag={tag} />
          </li>
        ))}
      </ul>
    </>
  );
}

export default CategoryPage;
