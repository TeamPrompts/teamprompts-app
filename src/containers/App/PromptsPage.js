import React from 'react';
import { useParams } from 'react-router-dom';
import Filters from '../../components/Filters/Filters';
import FillInTheBlanks from '../../FillInTheBlanks/FillInTheBlanks';
import Heading from './Heading';
import tagBySlug from './tagBySlug';
import { pageTypes } from '../../constants';

function PromptsPage({ fitbs, history, tags }) {
  const { id, slug } = useParams();
  const tag = tagBySlug({ slug, tags });
  const fitb = fitbs.find(fitb => fitb.id === id);
  return (
    <>
      <Heading title="TeamPrompts" />
      <Filters filter={{ name: tag.name }} fitbs={fitbs} tags={tags} />
      <ul>
        {fitb &&
          [fitb].map(fitb => (
            <li key={fitb.id}>
              <FillInTheBlanks
                fitb={fitb}
                history={history}
                pageType={pageTypes.PromptsPage}
                tag={tag}
              />
            </li>
          ))}
      </ul>
    </>
  );
}

export default PromptsPage;
