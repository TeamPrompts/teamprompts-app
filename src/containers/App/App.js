import React, { useEffect, useState } from 'react';
import { HashRouter as Router, Route, Switch } from 'react-router-dom';
import { version } from '../../../package.json';
import selectFitbs from '../../api/selectFitbs';
import selectTags from '../../api/selectTags';
import Footer from '../../components/Footer/Footer';
import Heading from './Heading';
import CategoryPage from './CategoryPage';
import HomePage from './HomePage';
import PromptsPage from './PromptsPage';

// TODO: change to _Provider
function Wrapper({ callback, children, loading }) {
  const [error, setError] = useState();
  const [fitbs, setFitbs] = useState([]);
  const [tags, setTags] = useState([]);
  const [waiting, setWaiting] = useState(true);

  useEffect(() => {
    selectFitbs((error, fitbs) => {
      if (error) {
        setError(error);
      } else {
        setFitbs(fitbs);
      }
    });
  }, []);

  useEffect(() => {
    selectTags((error, tags) => {
      if (error) {
        setError(error);
      } else {
        setTags(tags);
      }
    });
  }, []);

  useEffect(() => {
    if (error) {
      setWaiting(false);
    }
    if (fitbs.length > 0 && tags.length > 0) {
      setWaiting(false);
    }
  }, [error, fitbs.length, tags.length]);

  useEffect(() => {
    if (!waiting) {
      callback(error, { fitbs, tags });
    }
  }, [callback, error, fitbs, tags, waiting]);

  return waiting ? loading : children;
}

function App() {
  const [error, setError] = useState();
  const [fitbs, setFitbs] = useState([]);
  const [tags, setTags] = useState([]);

  return (
    <div className="flex flex-col font-serif items-center max-w-4xl mx-16 sm:mx-32 md:mx-32 lg:mx-32 xl:mx-auto">
      <Wrapper
        callback={(error, { fitbs, tags }) => {
          if (error) {
            setError(error);
          } else {
            setFitbs(fitbs);
            setTags(tags);
          }
        }}
        loading={
          <>
            <div className="mt-2">&nbsp;</div>
            <Heading title="Loading ..." />
          </>
        }
      >
        <Router>
          <Switch>
            <Route
              children={<PromptsPage fitbs={fitbs} tags={tags} />}
              exact={true}
              path="/:slug/:id"
            />
            <Route
              children={<CategoryPage fitbs={fitbs} tags={tags} />}
              exact={true}
              path="/:slug"
            />
            <Route
              children={<HomePage fitbs={fitbs} tags={tags} />}
              default={true}
              path="/"
            />
          </Switch>
        </Router>
      </Wrapper>
      <Footer version={version} />
    </div>
  );
}

export default App;
