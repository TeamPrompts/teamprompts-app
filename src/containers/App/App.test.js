import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';
import App from './App';
import selectFitbs from '../../api/selectFitbs';
import selectTags from '../../api/selectTags';
import fitb from '../../mocks/fitb';
import tag from '../../mocks/tag';

jest.mock('../../api/selectFitbs');
jest.mock('../../api/selectTags');

describe('App', () => {
  beforeEach(() => {
    selectFitbs.mockImplementation(callback => callback(null, [fitb]));
    selectTags.mockImplementation(callback => callback(null, [tag]));
  });

  it('loading', () => {
    const tree = create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('w/ fitbs & tags', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('w/ fitbs error', () => {
    selectFitbs.mockImplementation(callback => callback(new Error('Oh Noes!')));
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('w/ tags error', () => {
    selectTags.mockImplementation(callback => callback(new Error('Oh Noes!')));
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('click retrospectives', () => {
    const { getByTestId } = render(<App />);
    fireEvent.click(getByTestId(tag.name));
    expect(getByTestId(tag.name)).toMatchSnapshot();
  });

  // TODO: test routes
});
