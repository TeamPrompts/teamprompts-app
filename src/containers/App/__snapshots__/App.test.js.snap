// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App click retrospectives 1`] = `
<button
  class="capitalize focus:outline-none hover:bg-blue-700 px-4 rounded-full text-white bg-blue-900"
  data-testid="retrospectives"
  type="button"
>
  retrospectives (1)
</button>
`;

exports[`App loading 1`] = `
<div
  className="flex flex-col font-serif items-center max-w-4xl mx-16 sm:mx-32 md:mx-32 lg:mx-32 xl:mx-auto"
>
  <h1
    className="font-bold mb-8 mt-2 text-4xl text-gray-900"
  >
    Loading ...
  </h1>
  <p
    className="my-8"
  >
    <small
      className="font-normal text-gray-700 text-sm"
    >
      v
      0.16.19
    </small>
  </p>
</div>
`;

exports[`App w/ fitbs & tags 1`] = `
<div>
  <div
    class="flex flex-col font-serif items-center max-w-4xl mx-16 sm:mx-32 md:mx-32 lg:mx-32 xl:mx-auto"
  >
    <h1
      class="font-bold mb-8 mt-2 text-4xl text-gray-900"
    >
      TeamPrompts
    </h1>
    <ul
      class="flex flex-wrap mb-4"
    >
      <li
        class="mx-1 my-1"
      >
        <a
          href="#/all"
        >
          <button
            class="capitalize focus:outline-none hover:bg-blue-700 px-4 rounded-full text-white bg-blue-900"
            data-testid="all"
            type="button"
          >
            all (1)
          </button>
        </a>
      </li>
      <li
        class="mx-1 my-1"
      >
        <a
          href="#/retrospectives"
        >
          <button
            class="capitalize focus:outline-none hover:bg-blue-700 px-4 rounded-full text-white bg-blue-500"
            data-testid="retrospectives"
            type="button"
          >
            retrospectives (1)
          </button>
        </a>
      </li>
    </ul>
    <ul>
      <li>
        <div
          class="border-b border-gray-500 leading-loose mb-2 px-4 hover:bg-gray-100 cursor-pointer py-8 shadow text-lg"
          data-testid="container"
        >
          Do you know the way to 
          <span
            class="inline-flex flex-col leading-none"
            data-testid="input-container"
          >
            <div
              class="self-center"
              style="display: inline-block;"
            >
              <input
                class=""
                data-testid="input"
                placeholder="________________"
                style="box-sizing: content-box; width: 2px;"
                value=""
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              >
                ________________
              </div>
            </div>
            <label
              class="self-center text-sm"
            >
              place
            </label>
          </span>
          ?
          <span
            class="inline-block ml-3 relative"
          >
            <input
              class="hidden"
              data-testid="receM4Gus7HH43oLM"
              id="receM4Gus7HH43oLM"
              type="checkbox"
            />
            <label
              class="cursor-pointer"
              for="receM4Gus7HH43oLM"
            >
              <div
                class="w-6 h-3 bg-gray-400 rounded-full shadow-inner"
              />
              <div
                class="toggle__dot absolute w-4 h-4 bg-gray-300 rounded-full shadow inset-y-0 left-0"
              />
            </label>
          </span>
          <button
            class="bg-blue-500 focus:outline-none font-bold font-mono hover:bg-blue-700 ml-3 px-4 rounded-full text-xs text-white"
            data-testid="copy-button"
            type="button"
          >
            Copy to clipboard
          </button>
        </div>
      </li>
    </ul>
    <p
      class="my-8"
    >
      <small
        class="font-normal text-gray-700 text-sm"
      >
        v
        0.16.19
      </small>
    </p>
  </div>
</div>
`;

exports[`App w/ fitbs error 1`] = `
<div>
  <div
    class="flex flex-col font-serif items-center max-w-4xl mx-16 sm:mx-32 md:mx-32 lg:mx-32 xl:mx-auto"
  >
    <h1
      class="font-bold mb-8 mt-2 text-4xl text-gray-900"
    >
      TeamPrompts
    </h1>
    <ul
      class="flex flex-wrap mb-4"
    >
      <li
        class="mx-1 my-1"
      >
        <a
          href="#/all"
        >
          <button
            class="capitalize focus:outline-none hover:bg-blue-700 px-4 rounded-full text-white bg-blue-900"
            data-testid="all"
            type="button"
          >
            all (0)
          </button>
        </a>
      </li>
    </ul>
    <ul />
    <pre>
      {}
    </pre>
    <p
      class="my-8"
    >
      <small
        class="font-normal text-gray-700 text-sm"
      >
        v
        0.16.19
      </small>
    </p>
  </div>
</div>
`;

exports[`App w/ tags error 1`] = `
<div>
  <div
    class="flex flex-col font-serif items-center max-w-4xl mx-16 sm:mx-32 md:mx-32 lg:mx-32 xl:mx-auto"
  >
    <h1
      class="font-bold mb-8 mt-2 text-4xl text-gray-900"
    >
      TeamPrompts
    </h1>
    <ul
      class="flex flex-wrap mb-4"
    >
      <li
        class="mx-1 my-1"
      >
        <a
          href="#/all"
        >
          <button
            class="capitalize focus:outline-none hover:bg-blue-700 px-4 rounded-full text-white bg-blue-900"
            data-testid="all"
            type="button"
          >
            all (0)
          </button>
        </a>
      </li>
    </ul>
    <ul />
    <pre>
      {}
    </pre>
    <p
      class="my-8"
    >
      <small
        class="font-normal text-gray-700 text-sm"
      >
        v
        0.16.19
      </small>
    </p>
  </div>
</div>
`;
